README.md
<p align="center">
  <img src="https://raw.githubusercontent.com/Exelord/ember-initials/master/ember-initials.png" alt="ember-initials"/>
  <a href="https://travis-ci.org/Exelord/ember-initials">
    <img src="https://travis-ci.org/Exelord/ember-initials.svg?branch=master">
  </a>
  <a href="https://david-dm.org/exelord/ember-initials">
    <img src="https://david-dm.org/exelord/ember-initials/status.svg">
  </a>
  <a href="https://gitter.im/Exelord/ember-initials?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge">
    <img src="https://badges.gitter.im/Exelord/ember-initials.svg">
  </a>
  <a href="https://codeclimate.com/github/Exelord/ember-initials/maintainability">
    <img src="https://api.codeclimate.com/v1/badges/05c3ebd3e7a075ab21ef/maintainability">
  </a>
</p>

Ember Initials is a package for generating simple avatars using users initials, images, gravatars or adorables. It definitely supports Ember 2.12 (or higher). Thanks to highly customizable interface you can define defaults and even set fallbacks.

## Demo App with code generators
Demo app with interactive examples you can find at [http://exelord.github.io/ember-initials][9225812a]

[9225812a]: http://exelord.github.io/ember-initials/ "Ember Initials Demo App"

## Install
`ember install ember-initials`

## Quick start
Ember Initials gives you a built-in components which you can use in yours templates:

##### Initials avatars:
```hbs
{{ember-initials name="John Doe"}}
```

##### Image avatars:
```hbs
{{ember-initials/image image="/url/to/avatar.png"}}
```

##### Gravatar avatars:
```hbs
{{ember-initials/gravatar email="gravatar@email.com"}}
```

##### Adorables avatars:
```hbs
{{ember-initials/adorable email="email.to@generate.com"}}
```

> Examples, configuration and code generator are available on [Project Site][9225812a].

## API Documentation

### `ember-initials`
```hbs
{{ember-initials
  image="images/user.jpg"
  name="Ember Initials"
  defaultName="?"
  seedText="Ember Initials"
  textColor="white"
  fontSize=55
  fontWeight=100
  fontFamily="Helvetica Neue Light, Arial, sans-serif"
  size=200
}}
```

- `image` - image url | if image will be not found the initials avatars will be used

- `name` - name of user to generate initials

- `defaultName` - if `name` will be not passed this name will be used

- `seedText` - uniq string which will be used to generate background color, eg. user's email

- `textColor` - color of the text (initials) | hash or name

- `fontSize` - font size in `px`

- `fontWeight` - font weight in `px`

- `fontFamily` - names of fonts with fallbacks | at this time only system fonts can be used

- `size` - size of the avatar in pixels (set `null` if you do not want to use elements attributes like `width` and `height`)

### `ember-initials/image`
```hbs
{{ember-initials/image
  image="images/user.jpg"
  size=200
  defaultImage="images/default.jpg"
}}
```

- `image` - image url | if image will be not found the `defaultImage` will be used

- `defaultImage` - default image url

- `size` - size of the avatar in pixels (set `null` if you do not want to use elements attributes like `width` and `height`)

### `ember-initials/gravatar`
```hbs
{{ember-initials/gravatar
  image="images/user.jpg"
  email="example@example.com"
  defaultImage="images/default.jpg"
  relativeUrl="true"
  size=200
}}
```

- `image` - image url | if image will be not passed the avatar will be generated form email thanks to gravatar

- `email` - Email which will be used to fetching the gravatar

- `defaultImage` - image src (relative or absolute URL depends on `relativeUrl` option) | if gravatar will not be found this image will be used

- `relativeUrl` - `true` or `false` | decide which `defaultImage` url do you want to use - relative or absolute

- `size` - size of the avatar in pixels (set `null` if you do not want to use elements attributes like `width` and `height`)

### `ember-initials/adorable`
```hbs
{{ember-initials/adorable
  image="images/user.jpg"
  email="email@user.com"
  size=50
}}
```
- `image` - image url | if image will be not found the avatar will be generated form email thanks to adorable

- `email` - Email which will be used to generating unique adorable avatar

- `size` - size of the avatar in pixels (set `null` if you do not want to use elements attributes like `width` and `height`)

## Browser compatibility

This addon has been checked on:
- IE 10 and UP including EDGE
- Firefox 44 and UP
- Chrome 48 and UP
- Opera 35 and UP


CHANGELOG.md
# Change Log

## [v3.13.0](https://github.com/Exelord/ember-initials/tree/v3.13.0) (2018-07-04)
[Full Changelog](https://github.com/Exelord/ember-initials/compare/v3.12.2...v3.13.0)

**Implemented enhancements:**

- Update dependencies [\#66](https://github.com/Exelord/ember-initials/pull/66) ([Exelord](https://github.com/Exelord))

**Closed issues:**

- avatar using all width and height [\#65](https://github.com/Exelord/ember-initials/issues/65)
- Make alt text and title configurable [\#64](https://github.com/Exelord/ember-initials/issues/64)
- ember-initials is not a helper [\#58](https://github.com/Exelord/ember-initials/issues/58)

## [v3.12.2](https://github.com/Exelord/ember-initials/tree/v3.12.2) (2018-05-17)
[Full Changelog](https://github.com/Exelord/ember-initials/compare/v3.12.1...v3.12.2)

**Merged pull requests:**

- \[FIX\] Support for engine add-on [\#63](https://github.com/Exelord/ember-initials/pull/63) ([marcemira](https://github.com/marcemira))

## [v3.12.1](https://github.com/Exelord/ember-initials/tree/v3.12.1) (2018-05-06)
[Full Changelog](https://github.com/Exelord/ember-initials/compare/v3.12.0...v3.12.1)

**Closed issues:**

- Node.js Version Issues [\#60](https://github.com/Exelord/ember-initials/issues/60)

**Merged pull requests:**

- update engines package.json [\#61](https://github.com/Exelord/ember-initials/pull/61) ([donaldwasserman](https://github.com/donaldwasserman))

## [v3.12.0](https://github.com/Exelord/ember-initials/tree/v3.12.0) (2018-05-01)
[Full Changelog](https://github.com/Exelord/ember-initials/compare/v3.11.1...v3.12.0)

**Implemented enhancements:**

- Feature upgrades 3.1 [\#59](https://github.com/Exelord/ember-initials/pull/59) ([Exelord](https://github.com/Exelord))

**Closed issues:**

- Build Canceled: Broccoli Builder ran into an error with `Concat` plugin [\#55](https://github.com/Exelord/ember-initials/issues/55)

## [v3.11.1](https://github.com/Exelord/ember-initials/tree/v3.11.1) (2018-03-21)
[Full Changelog](https://github.com/Exelord/ember-initials/compare/v3.11.0...v3.11.1)

**Closed issues:**

- md5 issue with fresh install [\#56](https://github.com/Exelord/ember-initials/issues/56)

**Merged pull requests:**

- Fix md5 [\#57](https://github.com/Exelord/ember-initials/pull/57) ([Exelord](https://github.com/Exelord))

## [v3.11.0](https://github.com/Exelord/ember-initials/tree/v3.11.0) (2018-02-18)
[Full Changelog](https://github.com/Exelord/ember-initials/compare/v3.10.0...v3.11.0)

**Implemented enhancements:**

- Upgrade to ember 3.0 [\#54](https://github.com/Exelord/ember-initials/pull/54) ([Exelord](https://github.com/Exelord))

## [v3.10.0](https://github.com/Exelord/ember-initials/tree/v3.10.0) (2018-02-18)
[Full Changelog](https://github.com/Exelord/ember-initials/compare/v3.9.0...v3.10.0)

**Implemented enhancements:**

- Simplify build process [\#53](https://github.com/Exelord/ember-initials/pull/53) ([Exelord](https://github.com/Exelord))

## [v3.9.0](https://github.com/Exelord/ember-initials/tree/v3.9.0) (2018-02-05)
[Full Changelog](https://github.com/Exelord/ember-initials/compare/v3.8.0...v3.9.0)

**Implemented enhancements:**

- Remove object-hash dependency [\#52](https://github.com/Exelord/ember-initials/pull/52) ([Exelord](https://github.com/Exelord))

## [v3.8.0](https://github.com/Exelord/ember-initials/tree/v3.8.0) (2018-02-04)
[Full Changelog](https://github.com/Exelord/ember-initials/compare/v3.7.0...v3.8.0)

**Implemented enhancements:**

- Remove jQuery dependency [\#51](https://github.com/Exelord/ember-initials/pull/51) ([Exelord](https://github.com/Exelord))

## [v3.7.0](https://github.com/Exelord/ember-initials/tree/v3.7.0) (2018-01-19)
[Full Changelog](https://github.com/Exelord/ember-initials/compare/v3.6.1...v3.7.0)

**Implemented enhancements:**

- Upgrade Ember 2.18 [\#47](https://github.com/Exelord/ember-initials/pull/47) ([Exelord](https://github.com/Exelord))
- Add tests for object hash [\#44](https://github.com/Exelord/ember-initials/pull/44) ([Exelord](https://github.com/Exelord))

**Fixed bugs:**

- Newest release fails in fastboot [\#42](https://github.com/Exelord/ember-initials/issues/42)
- Allow to override colors and styles [\#50](https://github.com/Exelord/ember-initials/pull/50) ([Exelord](https://github.com/Exelord))

**Closed issues:**

- Cannot read property 'initialsFor' of undefined [\#46](https://github.com/Exelord/ember-initials/issues/46)

## [v3.6.1](https://github.com/Exelord/ember-initials/tree/v3.6.1) (2018-01-05)
[Full Changelog](https://github.com/Exelord/ember-initials/compare/v3.6.0...v3.6.1)

**Implemented enhancements:**

- Initialize cache store only when used [\#49](https://github.com/Exelord/ember-initials/pull/49) ([Exelord](https://github.com/Exelord))

**Fixed bugs:**

- Initialize cache store only when used [\#49](https://github.com/Exelord/ember-initials/pull/49) ([Exelord](https://github.com/Exelord))
- Fix container initializer [\#45](https://github.com/Exelord/ember-initials/pull/45) ([Exelord](https://github.com/Exelord))

## [v3.6.0](https://github.com/Exelord/ember-initials/tree/v3.6.0) (2017-12-26)
[Full Changelog](https://github.com/Exelord/ember-initials/compare/v3.5.0...v3.6.0)

**Implemented enhancements:**

- Feature refactor [\#41](https://github.com/Exelord/ember-initials/pull/41) ([Exelord](https://github.com/Exelord))

## [v3.5.0](https://github.com/Exelord/ember-initials/tree/v3.5.0) (2017-12-26)
[Full Changelog](https://github.com/Exelord/ember-initials/compare/v3.4.0...v3.5.0)

**Implemented enhancements:**

- Ember 2.17 upgrade [\#40](https://github.com/Exelord/ember-initials/pull/40) ([Exelord](https://github.com/Exelord))

## [v3.4.0](https://github.com/Exelord/ember-initials/tree/v3.4.0) (2017-11-27)
[Full Changelog](https://github.com/Exelord/ember-initials/compare/v3.3.0...v3.4.0)

**Implemented enhancements:**

- Packages upgrades [\#39](https://github.com/Exelord/ember-initials/pull/39) ([Exelord](https://github.com/Exelord))


addon/mixin.js
import Ember from 'ember';
import { colorIndex } from './-private/utils';
import { generateInitials, generateImage, revokeImage } from './-private/generators';

export default Ember.Mixin.create({
  tagName: 'img',
  attributeBindings: ['width', 'height', 'src'],

  defaultName: '?',
  defaultBackground: '#dd6a58',

  name: Ember.computed.reads('defaultName'),
  seedText: Ember.computed.reads('name'),

  size: 30,
  height: Ember.computed.reads('size'),
  width: Ember.computed.reads('size'),

  backgroundStyles: {},
  textStyles: {},

  textColor: 'white',
  fontSize: 50,
  fontWeight: 200,
  fontFamily: 'Helvetica Neue Light, Arial, sans-serif',

  colors: [
    '#1abc9c', '#16a085', '#f1c40f',
    '#f39c12', '#2ecc71', '#27ae60',
    '#e67e22', '#d35400', '#3498db',
    '#2980b9', '#e74c3c', '#c0392b',
    '#9b59b6', '#8e44ad', '#bdc3c7',
    '#34495e', '#2c3e50', '#95a5a6',
    '#7f8c8d', '#ec87bf', '#d870ad',
    '#f69785', '#9ba37e', '#b49255',
    '#b49255', '#a94136', '#5461b4',
  ],

  initialsObserver: Ember.observer('name', 'seedText', 'fontSize', 'fontWeight', 'fontFamily', 'textColor', 'defaultName', function () {
    revokeImage(this.get('src'));
    this.notifyPropertyChange('src');
  }),

  backgroundColor: Ember.computed('colors.length', 'seedText', 'defaultName', 'defaultBackground', function () {
    if (this.get('seedText') === this.get('defaultName')) {
      return this.get('defaultBackground');
    } else {
      let index = colorIndex(this.get('seedText'), this.get('colors.length'));
      return this.get('colors')[index];
    }
  }),

  src: Ember.computed('fastboot.isFastBoot', function() {
    if (!this.get('fastboot.isFastBoot')) {
      return this.createInitials();
    }
  }),

  fastboot: Ember.computed(function() {
    return Ember.getOwner(this).lookup('service:fastboot');
  }),

  createInitials() {
    return generateImage(this.initialsProperties());
  },

  initialsProperties() {
    return {
      width: 100,
      height: 100,
      initials: generateInitials(this.get('name') || this.get('defaultName')),
      initialsColor: this.get('textColor'),
      textStyles: Ember.assign(this._textStyles(), this.get('textStyles')),
      backgroundStyles: Ember.assign(this._backgroundStyles(), this.get('backgroundStyles')),
    };
  },

  willDestroyElement() {
    this._super(...arguments);
    revokeImage(this.get('src'));
  },

  _textStyles() {
    return {
      'font-family': this.get('fontFamily'),
      'font-weight': this.get('fontWeight'),
      'font-size': `${this.get('fontSize')}px`,
    };
  },

  _backgroundStyles() {
    return {
      'background-color': this.get('backgroundColor'),
    };
  },
});
